### PR TITLE
Remove option for fork-based compile pool

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142001

Summary: This has been set to "subproc" for a while internally and externally, so we can remove and simplify some of the code. Note that there's no pressing need here -- just that since we've had internal outage with the legacy "fork" implementation, it doesn't seem helpful to leave it available. But if people aren't in the mood for this sort of cleanup, I won't be offended to abandon it.

Test Plan: CI

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/inductor/test_codecache.py [hunk 1]
```diff
@@ -57,37 +56,6 @@
...
-
-        model = MyModel().to(device=GPU_TYPE)
-        model = torch.compile(model)
-        inp = torch.rand(10, 10).to(device=GPU_TYPE)
-        model(inp).sum().backward()
-
-
-@requires_gpu()
-def test_codecache_spawn():
-    _run_codecache_test("spawn")
```
Comment (masnesral) [inline test/inductor/test_codecache.py:83]: not sure what these were supposed to be doing, but they weren't running

File: torch/_inductor/compile_worker/subproc_pool.py [hunk 2]
```diff
@@ -306,7 +298,6 @@ def _warm_process_pool(pool: AnyPool, n: int) -> None:
 
     # We force them to start here with some YOLOing of the internal methods.
 
-    # TODO(masnesral): Are these still relevant?
```
Comment (masnesral) [inline torch/_inductor/compile_worker/subproc_pool.py:309]: Turns out this does speed up initialization of the ProcessPoolExecutor we use _inside_ the SubprocPool

**Review by eellison — APPROVED** [review]
🚀

**Review by jansel — APPROVED** [review]

**Comment by haifeng-jin** [comment]
@masnesral would you mind share a little bit more about how the usage of multiprocessing lead to a multiplicative number of compile subprocesses/"internal outage"?

We are still using the fork/spawn methods.
This PR blocks us from updating torch.
With your insights, we may be able to find a solution.

Thank you!

**Comment by masnesral** [comment]
@haifeng-jin Can you explain why you think you're usage is still dependent on "fork"? We've swapped out the previous implementation that created subprocesses by forking the main process with one that Popens a subprocess instead. That subprocess manages its own pool internally. It should be transparent except that the new approach avoids the common pitfalls of fork. 

The internal usage that lead to a large number of compile subprocesses basically boils down to spinning up subprocess before importing torch. Then each of those subprocesses spin up their own subprocess pool and we can end up with, e.g., 32x32 compile subprocesses. But that problem is orthogonal to this PR. That explosion of subprocesses could still happen. The way we're currently working around it is here: https://github.com/pytorch/pytorch/blob/8e483654cb58d16d18a69ae58df50d38d5930d28/torch/_inductor/async_compile.py#L324. Basically, we start the subprocess pool lazily and only spin it up on the first attempt to compile Triton.

**Comment by kiukchung** [comment]
> @haifeng-jin Can you explain why you think you're usage is still dependent on "fork"? We've swapped out the previous implementation that created subprocesses by forking the main process with one that Popens a subprocess instead. That subprocess manages its own pool internally. It should be transparent except that the new approach avoids the common pitfalls of fork.
> 
> The internal usage that lead to a large number of compile subprocesses basically boils down to spinning up subprocess before importing torch. Then each of those subprocesses spin up their own subprocess pool and we can end up with, e.g., 32x32 compile subprocesses. But that problem is orthogonal to this PR. That explosion of subprocesses could still happen. The way we're currently working around it is here:
> 
> https://github.com/pytorch/pytorch/blob/8e483654cb58d16d18a69ae58df50d38d5930d28/torch/_inductor/async_compile.py#L324
> 
> . Basically, we start the subprocess pool lazily and only spin it up on the first attempt to compile Triton.

Hi @masnesral thanks for the explanation. The issue we (at Google) are having is that:

1. Its tricky for us to use `subprocess` in Borg deployments due to the fact that Python binaries/libraries are packaged as a hermetic executable (e.g. think PAR files). This is technically possible by creating a `py_binary()` Bazel target for [`torch._inductor.compile_worker.__main__`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_worker/__main__.py) but since `compile_worker.__main__` has a dep to `torch` it means we'd be deploying two copies of `torch`.

2. prior to this change we had an option to use `multiprocessing` (over `subprocess`). Unfortunately Google's codebase is not `fork`-safe so we were using `forkserver` or `spawn` for the compiler worker pool (which has been working out for us). The reason for the question was to see if there was something we missed that inherently makes it a bad idea to use `multiprocessing` (over subprocess). We could assess the edge-cases and figure out whether we can continue using `forkserver`/`spawn`.

**Comment by haifeng-jin** [comment]
Hi @masnesral,

Thank you so much for your answer!
As explained by @kiukchung, I wonder if we could keep the option for users to use `forkserver` or `spawn`.
It doesn't have to be an exact revert of this PR.
As long as it allows us to use `multiprocessing`, it should be good enough.
If it sounds good to you, I can work on this.

What do you suggest for the change? Thanks!

**Comment by masnesral** [comment]
> Its tricky for us to use subprocess in Borg deployments

@kiukchung @haifeng-jin I see, yeah that makes sense. Sorry, I didn't consider anyone to have been using anything other than "fork" or "subprocess" because I thought we've had an assert in place for a while. But now I see the assert is only in `config.py`, so it wouldn't have triggered if you're setting programmatically. 

Lemme put up a PR to restore the functionality. As you anticipated, it'll be a bit different than a straight revert. Perhaps you can help me test it?

**Comment by haifeng-jin** [comment]
Thank you, @masnesral !

I am happy to test it out when you have the PR.

**Comment by masnesral** [comment]
> I am happy to test it out when you have the PR.

@haifeng-jin: https://github.com/pytorch/pytorch/pull/144491

**Pull request closed without merging**